### PR TITLE
feat: push notifications new controller

### PR DESCRIPTION
**Pull request by matteoscurati**
## **Description**

This PR introduces a new SW (Service Worker) and a new controller for managing push notifications.

This integration, due to the differences between MV2 and MV3 regarding the use of background scripts and the SW, is only related to the manifest V2. For the manifest V3, part of this logic will be modified, using the extension's own service worker as a starting point. With manifest V2, on the contrary, it is necessary to initialize a dedicated service worker so that FCM (Firebase Cloud Messaging) can handle messages received in the background.

The final UI will allow the user to enable/disable the receipt of notifications. However, this PR concerns only the

new controller and the build management for the creation of the SW.

The new controller will work in accordance with two other controllers for managing authentication and for managing the triggers necessary to receive notifications.

## **Related issues**

No related issues.

Fixes:

## **Manual testing steps**

At the moment, the necessary environment variables have not been inserted into the build process. To test the branch locally, feel free to ask me for the env to insert into your local .metamaskrc file

## **Screenshots/Recordings**

In this video, I show the interaction between a test server for sending notifications and the extension:

https://www.loom.com/share/702fc6a8aea74510acdb1bfac614fa4a

### **After**

Check the entire flow here https://www.figma.com/file/c7GgNw2kScGrVyRGAPhwEd/Notifications?type=design&node-id=484%3A44713&mode=design&t=QsKrmdqZnx42IrMt-1

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've clearly explained what problem this PR is solving and how it is solved.
- [ ] I've linked related issues
- [ ] I've included manual testing steps
- [x] I've included screenshots/recordings if applicable
- [ ] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - [x] In case it's not yet "ready for review", I've set it to "draft".
  - [ ] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: app/scripts/controllers/push-platform-notifications/firebase/lib/firebase-app.js
Comment (danjm) [inline app/scripts/controllers/push-platform-notifications/firebase/lib/firebase-app.js:1]: @matteoscurati could you include a link to where you copied the code from in this code comment?

File: app/scripts/controllers/push-platform-notifications/firebase/lib/firebase-messaging-sw.js
Comment (danjm) [inline app/scripts/controllers/push-platform-notifications/firebase/lib/firebase-messaging-sw.js:1]: A few questions:
- why does this file need to be minified?
- is it copied from somewhere? if so where? and then does a license need to be included?
- can we include a comment at the top of this file explaining what it is?

Comment (danjm) [reply app/scripts/controllers/push-platform-notifications/firebase/lib/firebase-messaging-sw.js:1]: Same for the below file

File: app/scripts/controllers/push-platform-notifications/firebase/utils/get-notification-data.ts
Comment (danjm) [inline app/scripts/controllers/push-platform-notifications/firebase/utils/get-notification-data.ts:14]: I see you added a code comment for the same function in shared/lib/transactions-controller-utils.js. Did you intend to import that function here?

File: shared/constants/platform-notifications.ts
Comment (danjm) [inline shared/constants/platform-notifications.ts:141]: This const seems to serve a similar purpose as the `CURRENCY_SYMBOLS` or `CHAINLIST_CURRENCY_SYMBOLS_MAP` const in `shared/constants/network.ts`. However, some of the symbols here are different from the symbols in that file (for example, the optimism symbol is different).

I might make sense to import the symbols from `shared/constants/network.ts` to this file, to make sure we are using the same symbols throughout the application.

(Which are the correct symbols is a separate question, but we should make sure we are consistent across the application)

File: shared/constants/platform-notifications.ts
Comment (danjm) [inline shared/constants/platform-notifications.ts:22]: Nit: because we refer to `ETHEREUM` as `MAINNET` throughout the `shared/constants/network` file, and that is used throughout the codebase, it might be best to name this `chains.ETHEREUM` property as `chains.MAINNET`

File: app/scripts/controllers/push-platform-notifications/firebase/utils/get-notification-message.ts
Comment (danjm) [inline app/scripts/controllers/push-platform-notifications/firebase/utils/get-notification-message.ts:83]: These fallbacks to english strings (e.g. `?? 'Funds sent'`) should not be needed, because the `t` function from the `translate.ts` file falls back to the english translation if no translation in the current locale exists

Comment (danjm) [reply app/scripts/controllers/push-platform-notifications/firebase/utils/get-notification-message.ts:83]: the same applied to all cases in this file

File: app/scripts/controllers/push-platform-notifications/push-platform-notifications.ts
Comment (danjm) [inline app/scripts/controllers/push-platform-notifications/push-platform-notifications.ts:48]: Could you add a brief code comment that describes the purpose of the controller?

File: app/scripts/controllers/push-platform-notifications/types/on-chain-notification/schema.d.ts
Comment (danjm) [inline app/scripts/controllers/push-platform-notifications/types/on-chain-notification/schema.d.ts:2]: Perhaps include one more line here on how to generate the file, maybe even just the specific command that was used?

File: app/scripts/controllers/push-platform-notifications/push-platform-notifications.ts
Comment (danjm) [inline app/scripts/controllers/push-platform-notifications/push-platform-notifications.ts:104]: Any reason not to include the `log.error` message in the new Error thrown below? I think that with the current implementation, Sentry won't capture the specific error message.

**Review by danjm — COMMENTED** [review]
Okay, I have completed a first review of this PR. Looks really good overall. My comments are relatively minor

**Review by danjm — CHANGES_REQUESTED** [review]
Okay, I have completed a first review of this PR. Looks really good overall. My comments are relatively minor

File: app/scripts/controllers/push-platform-notifications/firebase/lib/firebase-app.js
Comment (matteoscurati) [reply app/scripts/controllers/push-platform-notifications/firebase/lib/firebase-app.js:1]: You're right. I'll include the license and the source of these libraries 👍

File: app/scripts/controllers/push-platform-notifications/firebase/lib/firebase-messaging-sw.js
Comment (matteoscurati) [reply app/scripts/controllers/push-platform-notifications/firebase/lib/firebase-messaging-sw.js:1]: Ok, see the comment above.


The library is minified because it is published that way here https://www.gstatic.com/firebasejs/10.8.0/firebase-messaging-sw.js.

In any case, I have proceeded to expand it.

Comment (matteoscurati) [reply app/scripts/controllers/push-platform-notifications/firebase/lib/firebase-messaging-sw.js:1]: Sure. See the comment  above.

The library is minified because it is published that way here [https://www.gstatic.com/firebasejs/10.8.0/firebase-messaging-sw.js](https://www.gstatic.com/firebasejs/10.8.0/firebase-messaging-sw.js).

In any case, I have proceeded to expand it.

File: shared/constants/platform-notifications.ts
Comment (matteoscurati) [reply shared/constants/platform-notifications.ts:141]: Yes, you're absolutely right. I've updated the file!

File: app/scripts/controllers/push-platform-notifications/firebase/utils/get-notification-message.ts
Comment (matteoscurati) [reply app/scripts/controllers/push-platform-notifications/firebase/utils/get-notification-message.ts:83]: Ah ok, perfect!

Comment (matteoscurati) [reply app/scripts/controllers/push-platform-notifications/firebase/utils/get-notification-message.ts:83]: Uhm, now I remember why I had included this fallback. Without it, I get an eslint error "Type 'string | null' is not assignable to type 'string'". I'll try to investigate a bit more!

Comment (matteoscurati) [reply app/scripts/controllers/push-platform-notifications/firebase/utils/get-notification-message.ts:83]: Ok, solved!

File: app/scripts/controllers/push-platform-notifications/push-platform-notifications.ts
Comment (matteoscurati) [reply app/scripts/controllers/push-platform-notifications/push-platform-notifications.ts:104]: No, you're right. I've corrected it using this pattern:

```
      const errorMessage = `Failed to update triggers for push notifications: ${error}`;
      log.error(errorMessage);
      throw new Error(errorMessage);
```

Can it work?

**Comment by matteoscurati** [comment]
@danjm  thank you very much!

I've reviewed your comments and modified the code as requested. I hope everything is ok.

If so, I'll wait to follow the order for merging the PRs we discussed to:
- change the status of the PR from draft to ready to be merged
- fix the build tests

File: app/scripts/controllers/push-platform-notifications/firebase/lib/firebase-app.js
Comment (rekmarks) [reply app/scripts/controllers/push-platform-notifications/firebase/lib/firebase-app.js:1]: Can we not use the npm package for this?

File: app/scripts/controllers/push-platform-notifications/push-platform-notifications.ts
Comment (rekmarks) [inline app/scripts/controllers/push-platform-notifications/push-platform-notifications.ts:1]: Why is this controller implemented directly in the extension as opposed to `/core`?

File: app/scripts/controllers/push-platform-notifications/firebase/lib/firebase-app.js
Comment (matteoscurati) [reply app/scripts/controllers/push-platform-notifications/firebase/lib/firebase-app.js:1]: Great question. Unfortunately, Lavamoat node doesn't support ESM, and Firebase does not expose an entry point in CJS. For this reason, at least for now, we had to import Firebase in the way indicated. If I understand correctly, this problem will be solved with the release of Endomoat

File: app/scripts/controllers/push-platform-notifications/push-platform-notifications.ts
Comment (matteoscurati) [reply app/scripts/controllers/push-platform-notifications/push-platform-notifications.ts:1]: Yes, we've already discussed this issue for all the other controllers we're using to import the new feature dedicated to notifications. You're absolutely right, and ultimately, we want to implement this and the other controllers in core, but currently, for reasons of speed in implementation, we've decided to work directly in the extension

File: app/scripts/controllers/push-platform-notifications/firebase/lib/firebase-app.js
Comment (FrederikBolding) [reply app/scripts/controllers/push-platform-notifications/firebase/lib/firebase-app.js:1]: Can we transpile Firebase using Babel instead? 

There is an example of how we did this in the past here: https://github.com/MetaMask/metamask-extension/pull/17251/files#diff-a05d1f077953d24c0d8ddecf64163da2653ba3b728e1ba1f18af6d770767b127R831-R841

Comment (matteoscurati) [reply app/scripts/controllers/push-platform-notifications/firebase/lib/firebase-app.js]: Ok, thanks to @FrederikBolding 's suggestion, I'm able to import Firebase using npm instead of importing the libraries directly!

**Review by danjm — DISMISSED** [review]
All of my comments have been addressed. I am approving. We will need to get reviews from other codeowners.

**Review by danjm — DISMISSED** [review]

File: app/scripts/metamask-controller.js
Comment (danjm) [inline app/scripts/metamask-controller.js:2868]: I think we can avoid having to do this. If the `PushPlatformNotificationsController` is added to `this.store` in the metamask-controller but not `this.memStore`, then it's state won't be exposed to the UI, but it will still be persisted.

Comment (matteoscurati) [reply app/scripts/metamask-controller.js]: Ok, thanks for the comment, now everything is much clearer. Removed. Thx!

**Review by Prithpal-Sooriya — COMMENTED** [review]
Some comments - will resolve them

File: app/scripts/controllers/metamask-notifications/constants/notification-schema.ts
Comment (Prithpal-Sooriya) [inline app/scripts/controllers/metamask-notifications/constants/notification-schema.ts:44]: Hardcoded Chain Symbols since we need Platform agnostic values (not tied to extension).

If there are shared values then sweet, otherwise we'll keep as is for now.

File: shared/constants/platform-notifications.ts
Comment (Prithpal-Sooriya) [reply shared/constants/platform-notifications.ts]: I have remove this file :)

File: app/scripts/controllers/push-platform-notifications/firebase/utils/get-notification-message.ts
Comment (Prithpal-Sooriya) [reply app/scripts/controllers/push-platform-notifications/firebase/utils/get-notification-message.ts]: Yep nice callout!

File: lavamoat/browserify/beta/policy.json
Comment (Prithpal-Sooriya) [inline lavamoat/browserify/beta/policy.json:3751]: A good chunk of firebase permissions.

It sucks that the npm module is a "large" npm module, when realistically we only use App and Messaging.

These do get tree shaken, but guessing lavamoat policies run on the node modules?

**Review by Prithpal-Sooriya — APPROVED** [review]
LGTM! We will wait for an extension review (and also lets merge this after next RC to avoid additional MV3 debugging)

**Review by danjm — APPROVED** [review]
LGTM

**Pull request merged**
